magento/magento2
# Ability to connect to Redis Sentinel for Cache, PageCache and Session

### Description


This PR provides the support of [Redis Sentinel](https://redis.io/topics/sentinel) for Cache, PageCache, and Session.

Usage of Redis Sentinel is suitable in an HA environment.

This is supported "out of the box" by colinmollenhour modules (credis, php-abstract-session, and it's cache backend). See https://github.com/colinmollenhour/Cm_Cache_Backend_Redis#high-availability-and-load-balancing-support

But Magento is not actually fully compatible with Redis Sentinel usage. This PR intends to add this compatibility.

What has been done : 

- leveraging dependencies to colinmollenhour modules.
- adding the ability to support additional configuration variables related to Redis Sentinel when installing with command line.
- updating unit tests to remain compliant.

New variables added to the `setup:install` in command line : 

	--cache-backend-redis-sentinel-master
	--cache-backend-redis-sentinel-master-verify
	--cache-backend-redis-sentinel-load-from-slaves
	--page-cache-redis-sentinel-master
	--page-cache-redis-sentinel-master-verify
	--page-cache-redis-sentinel-load-from-slaves
	--session-save-redis-sentinel-master
	--session-save-redis-sentinel-master-verify
	--session-save-redis-sentinel-load-from-slaves
	--session-save-redis-sentinel-connect-retries

Basically, there are three parameters that are identical in the 3 cases : 

|Argument | Required ? | Values | Comment |
---------|-----------|-------------|----------|
| --cache-backend-redis-sentinel-master | No | string |the name of the Redis master which is monitored by the sentinel servers. |
| --cache-backend-redis-sentinel-master-verify | No | 0/1 |Verify connected server is actually master as per Sentinel client spec. |
| --cache-backend-redis-sentinel-load-from-slaves | No | 1/2 |A random slave will be chosen for performing reads in order to load balance across multiple Redis instances. Using the value '1' indicates to only load from slaves and '2' to include the master in the random read slave selection. |
| --page-cache-redis-sentinel-master | No | string |the name of the Redis master which is monitored by the sentinel servers. |
| --page-cache-redis-sentinel-master-verify | No | 0/1 |Verify connected server is actually master as per Sentinel client spec. |
| --page-cache-redis-sentinel-load-from-slaves | No | 1/2 |A random slave will be chosen for performing reads in order to load balance across multiple Redis instances. Using the value '1' indicates to only load from slaves and '2' to include the master in the random read slave selection. |
| --session-save-redis-sentinel-master | No | string |the name of the Redis master which is monitored by the sentinel servers. |
| --session-save-redis-sentinel-master-verify | No | 0/1 |Verify connected server is actually master as per Sentinel client spec. |
| --session-save-redis-sentinel-load-from-slaves | No | 1/2 |A random slave will be chosen for performing reads in order to load balance across multiple Redis instances. Using the value '1' indicates to only load from slaves and '2' to include the master in the random read slave selection. |
| --session-save-redis-sentinel-connect-retries | No | string |Number of connect retries when contacting sentinels  |

Please also note that when using Sentinel, you must use the IP/address of your sentinel servers as your Redis Host. Since you are meant to have several sentinels when running such an instance, you can use the list of all your sentinel servers as an input for the "host" or "server" configuration values for Redis (Eg : 'tcp://10.0.3.1:26379,tcp://10.0.3.2:26379,tcp://10.0.3.3:26379') . 

Here is an example of a generated `app/etc/env.php` : 

```
  'session' => 
  array (
    'save' => 'redis',
    'redis' => 
    array (
      'host' => 'tcp://10.0.3.1:26379,tcp://10.0.3.2:26379,tcp://10.0.3.3:26379', // Sentinel servers
      'port' => '26379',
      'password' => '',
      'timeout' => '2.5',
      'persistent_identifier' => '',
      'database' => '0',
      'compression_threshold' => '2048',
      'compression_library' => 'gzip',
      'log_level' => '1',
      'max_concurrency' => '6',
      'break_after_frontend' => '5',
      'break_after_adminhtml' => '30',
      'fail_after_frontend' => '1',
      'fail_after_adminhtml' => '10',
      'first_lifetime' => '600',
      'bot_first_lifetime' => '60',
      'bot_lifetime' => '7200',
      'disable_locking' => '0',
      'min_lifetime' => '60',
      'max_lifetime' => '2592000',
      'sentinel_master' => 'redismaster', // Name of master monitored by Sentinels.
      'sentinel_master_verify' => '1',
      'load_from_slaves' => '2',
      'sentinel_connect_retries' => '3',
    ),
  ),
  'cache' => 
  array (
    'frontend' => 
    array (
      'default' => 
      array (
        'backend' => 'Cm_Cache_Backend_Redis',
        'backend_options' => 
        array (
          'server' => 'tcp://10.0.3.1:26379,tcp://10.0.3.2:26379,tcp://10.0.3.3:26379', // Sentinel servers
          'database' => '1',
          'port' => '26379',
          'sentinel_master' => 'redismaster', // Name of master monitored by Sentinels.
          'sentinel_master_verify' => '1',
          'load_from_slaves' => '2',
        ),
      ),
      'page_cache' => 
      array (
        'backend' => 'Cm_Cache_Backend_Redis',
        'backend_options' => 
        array (
          'server' => 'tcp://10.0.3.1:26379,tcp://10.0.3.2:26379,tcp://10.0.3.3:26379', // Sentinel servers
          'database' => '2',
          'port' => '26379',
          'compress_data' => '0',
          'sentinel_master' => 'redismaster', // Name of master monitored by Sentinels.
          'sentinel_master_verify' => '1',
          'load_from_slaves' => '2',
        ),
      ),
    ),
  ),
```

### Manual testing scenarios

1. ... Install a Redis Sentinel Cluster somewhere (there are some "fast to setup" examples with Docker-Compose)
2. ... Install Magento2 with command line and specify these set of parameters : 

```
	--cache-backend=redis \
        --cache-backend-redis-server=tcp://10.0.3.1 \
        --cache-backend-redis-db=1 \
        --cache-backend-redis-port=26379 \
	--cache-backend-redis-sentinel-master=redismaster \
	--cache-backend-redis-sentinel-master-verify=1 \
	--cache-backend-redis-sentinel-load-from-slaves=2 \
        --page-cache=redis \
        --page-cache-redis-server=tcp://10.0.3.1:26379 \
        --page-cache-redis-db=2 \
        --page-cache-redis-port=26379 \
	--page-cache-redis-sentinel-master=redismaster \
	--page-cache-redis-sentinel-master-verify=1 \
	--page-cache-redis-sentinel-load-from-slaves=2 \
        --session-save=redis \
        --session-save-redis-host=tcp://10.0.3.1:26379 \
        --session-save-redis-db=0 \
	--session-save-redis-port=26379 \
	--session-save-redis-host=tcp://10.0.3.1:26379 \
	--session-save-redis-sentinel-master=redismaster \
	--session-save-redis-sentinel-master-verify=1 \
	--session-save-redis-sentinel-load-from-slaves=2 \
	--session-save-redis-sentinel-connect-retries=3 \
```

3. Check everything is running properly (cache keys are here in Redis, the same for page cache and sessions).


### Contribution checklist
